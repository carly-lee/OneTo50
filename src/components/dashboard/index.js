export { default as Dashboard } from './Dashboard';
export { default as Timer } from './Timer';
export { default as NextNumber } from './NextNumber';