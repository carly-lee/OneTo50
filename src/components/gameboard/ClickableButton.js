import React, { PropTypes } from 'react';
import { TouchableOpacity, Text } from 'react-native';

//@flow
export const ClickableButton = ( props )=>{
  const _onPress = ()=>{
    props.onPress();
  };

  return (
    <TouchableOpacity style={ props.style } onPress={ _onPress }>
      { props.child }
    </TouchableOpacity>
  )
};
 
ClickableButton.propTypes = {
  child: PropTypes.element.isRequired,
  onPress: PropTypes.func.isRequired,
}

