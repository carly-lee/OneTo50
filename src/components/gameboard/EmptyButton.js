// @flow
// $FlowIgnore
import { StyleSheet, View } from 'react-native';
import React from 'react';

export const EmptyButton = ()=>(
  <View style={styles.container}></View>
);

const styles = StyleSheet.create({
  container: {
    flex: 0,
    width: 70,
    height: 70,
    flexDirection: 'column',
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: 'white'
  }
});