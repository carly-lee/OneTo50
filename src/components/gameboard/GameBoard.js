//@flow
import React, { PureComponent, PropTypes } from 'react';
import { Text, View, StyleSheet } from 'react-native';

export default class GameBoard extends PureComponent{
  render(){
    return(
      <View style={ styles.container }>
        <Text>GameBoard</Text>
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
});