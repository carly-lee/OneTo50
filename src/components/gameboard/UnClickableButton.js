import React, { PropTypes } from 'react';
import { View, Text } from 'react-native';

//@flow
export const UnClickableButton = ({...props})=>{
  return (
    <View style={ props.style }>
      { props.child }
    </View>
  );
}

UnClickableButton.propTypes = {
  child: PropTypes.element.isRequired,
}
