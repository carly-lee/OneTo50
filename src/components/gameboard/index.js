export { default as PlayBoard } from './PlayBoard';
export { NumberButton } from './NumberButton';
export { EmptyButton } from './EmptyButton';
export { ClickableButton } from './ClickableButton';
export { UnClickableButton } from './UnClickableButton';