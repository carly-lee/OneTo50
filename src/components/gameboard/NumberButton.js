import React, { PureComponent } from 'react';
import { Text, StyleSheet } from 'react-native';
import { ClickableButton, UnClickableButton } from 'components/gameboard';

//@flow
type Props = {
  number: number,
  index: number,
  nextNumber: number,
  onClickNumber: ( num: number ) => void,
};

export default class NumberButton extends PureComponent{
  props: Props;

  _onPress = ():void =>{
    this.props.onClickNumber( this.props.number , this.props.index );
  }

  render(){
    const{ number, nextNumber } = this.props;
    const child:React.Element = <Text style={ styles.number }>{ this.props.number }</Text>;

    if( number === nextNumber ){
      return <ClickableButton style={styles.container} child={child} onPress={this._onPress} />;
    }else{
      return <UnClickableButton style={styles.container} child={child} />;
    }
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 0,
    width: 70,
    height: 70,
    flexDirection: 'column',
    justifyContent: 'center',
    alignItems: 'center',
    borderWidth: 1,
    borderColor: 'white',
  },
  number: {
    fontSize: 24,
    fontWeight: 'bold',
  },
});