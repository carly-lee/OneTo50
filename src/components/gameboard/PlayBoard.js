import React, { PureComponent } from 'react';
import { View, StyleSheet } from 'react-native';

import { NumberButton, EmptyButton } from 'components/gameboard';
import { shuffleArray } from 'utils';

//@flow
type Props = {
  nextNumber: number,
  onNumberClick: ( num: number ) => void,
};

export default class PlayBoard extends PureComponent{
  props:Props;
  allNumbers:Array<number> = [];

  state: {
    buttons:Array<NumberButton>,
  };

  constructor( props ){
    super( props );
    const initNumbers:Array<number> = this._getShuffledNumbers();
    let buttons:Array<NumberButton> = [];
    for( let i=0; i<25; i++ ){
      this.allNumbers.push( i+26 );
      buttons.push( <NumberButton key={ i } number={ initNumbers[i] } index={ i } nextNumber={ this.props.nextNumber } onClickNumber={ this._onClickNumber }  /> );
    }
    this.state = {
      buttons,
    };
  }

  _getShuffledNumbers = ():Array<number>=>{
    let array:Array<NumberButton> = [];
    for( let i=1; i<=25; i++ ){
      array.push( i );
    }
    return shuffleArray( array );
  }

  _onClickNumber = ( num:number, index:number ):void =>{
    const nextClick = num + 1;
    let newButtons:Array<NumberButton>;
    if( this.allNumbers.length > 0 ){
      newButtons = this._getButtons( index, nextClick );
    }else{
      newButtons = this._disappearButtons( index, nextClick );
    }
    this.setState({ buttons: newButtons });
    this.props.onNumberClick( nextClick );
  }

  _getButtons = ( index:number, nextClick:number ):Array<NumberButton>=>{
    const next:number = this.allNumbers.splice( 0,1 )[0];
    return this.state.buttons.map(( btn, idx )=>{
      if( idx === index ){
        return React.cloneElement( btn, { number: next, nextNumber: nextClick });
      }else{
        return React.cloneElement( btn, { nextNumber: nextClick });
      }
    });
  }

  _disappearButtons = ( index:number, nextClick:number )=>{
    return this.state.buttons.map(( btn, idx )=>{
      if( idx === index ){
        return <EmptyButton key={ idx } />;
      }else{
        return React.cloneElement( btn, { nextNumber: nextClick });
      }
    });
  }

  render(){
    return(
      <View style={ styles.container }>
      { this.state.buttons }
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 0,
    width: 350,
    height: 350,
    flexDirection: 'row',
    flexWrap: 'wrap',
    backgroundColor: 'grey',
  },
});