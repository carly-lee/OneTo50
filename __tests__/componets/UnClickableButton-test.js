import 'react-native';
import React from 'react';
import { Text } from 'react-native';
import { UnClickableButton } from 'components/gameboard';
import renderer from 'react-test-renderer';

describe( 'UnClickableButton', ()=>{
  const child = <Text>test</Text>;

  it( 'renders correctly',()=>{
    const tree = renderer.create( <UnClickableButton child={child} /> );
    expect( tree ).toMatchSnapshot();
  });
});