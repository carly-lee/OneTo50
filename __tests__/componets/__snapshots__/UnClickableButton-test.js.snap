exports[`UnClickableButton renders correctly 1`] = `
<View
  style={undefined}>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail">
    test
  </Text>
</View>
`;
